hiro/cocoa/widget/hex-edit.cpp

```diff
@@ -5,43 +5,167 @@
+-(id) tableView:(NSTableView *) tableView
+objectValueForTableColumn:(NSTableColumn *) tableColumn
+            row:(NSInteger) row {
+  // return a string with the content of (row, column)
+  // u32 address = hexEdit->address() + row * hexEdit->columns();
+  u32 address = row * hexEdit->columns();
+  // address only
+  if ([[tableColumn identifier] isEqualToString:@"Address"]) {
+    return [NSString stringWithUTF8String:hex(address, 8L).data()];
+  } else if ([[tableColumn identifier] isEqualToString:@"Char"]){
+    // create the char string by concatenating
+    NSMutableString *output = [NSMutableString stringWithCapacity:hexEdit->columns()];
+    for (auto column : range(hexEdit->columns())) {
+      if (address < hexEdit->length()) {
+        u8 data = hexEdit->doRead(address++);
+        [output appendString:[NSString stringWithFormat:@"%c",
+                              (data >= 0x20 && data <= 0x7e ? (char)data : '.')]];
+      }
+    }
+    return output;
+  } else {
+    // return only the single byte at the correct position
+    NSInteger columnNumber = [tableColumn.identifier integerValue];
+    address += columnNumber;
+    if (address < hexEdit->length()) {
+      u8 data = hexEdit->doRead(address);
+      return [NSString stringWithUTF8String:hex(data, 2L).data()];
+    } else {
+      return @"  ";
+    }
+  }
 }
 
-auto pHexEdit::setLength(u32 length) -> void {
+-(void) tableView:(NSTableView *) tableView
+   setObjectValue:(id) object
+   forTableColumn:(NSTableColumn *) tableColumn
+              row:(NSInteger) row {
+  // when table is edited, modify underlying data source
+  NSInteger colNumber = [tableColumn.identifier integerValue];
+  u32 address = row * hexEdit->columns() + colNumber;
+  
+  // unclear if this is needed, but just to make sure the cast is safe
+  if ([object isKindOfClass:[NSString class]]) {
+    if (address < hexEdit->length()) {
+      // only get the first 2 characters
+      NSString* newVal = [(NSString *)object substringToIndex:2];
+      NSScanner* hexScanner = [NSScanner scannerWithString:newVal];
+      unsigned int data = 0;
+      if ([hexScanner scanHexInt:&data]) {
+        // this.......is probably ok....???
+        hexEdit->doWrite(address, (u8)data);
+      }
+    }
+    
+  }
+  [tableView reloadData];
 }
 
-auto pHexEdit::setRows(u32 rows) -> void {
-}
+@end
 
-auto pHexEdit::update() -> void {
-}
+namespace hiro {
+  
+  auto pHexEdit::construct() -> void {
+    cocoaView = cocoaHexEdit = [[CocoaHexEdit alloc] initWith:self()];
+    pWidget::construct();
 
+    for (NSTableColumn *column in cocoaHexEdit.tableView.tableColumns) {
+        if (@available(macOS 10.15, *)) {
+        NSTextFieldCell *cell = column.dataCell;
+          NSTableHeaderCell *headerCell = column.headerCell;
+        
+        // Set the font for the data cell.
+        cell.font = [NSFont monospacedSystemFontOfSize:10 weight:NSFontWeightRegular];
+      }
+    }
```

Comment:
Just some messed up indentation.

```suggestion
    for (NSTableColumn *column in cocoaHexEdit.tableView.tableColumns) {
      if (@available(macOS 10.15, *)) {
        NSTextFieldCell *cell = column.dataCell;
        NSTableHeaderCell *headerCell = column.headerCell;
      
        // Set the font for the data cell.
        cell.font = [NSFont monospacedSystemFontOfSize:10 weight:NSFontWeightRegular];
      }
    }
```